[kaka0 Adult] ProductView 상품정보 조건에 맞게 조회 기능 추가

jotasic opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 상품 조회 기능 추가
		 
## 수정 사항들 자세한 내용
- 쿼리 스트링에 따라 상품 결과가 반환 되도록 기능 추가함.

/products?`type`=타입명&`keyword`=키워드&`pageSize`=페이지사이즈&`page`=페이지번호
- type : 필터링 타입 new,hot,search,category,character가 값으로 올 수 있음
    - new : 신상품 순 => keyword 필요없음
    - hot : 좋아요가 많은 수 => keyword 필요없음
    - search : 제품명으로 검색시 => keyword : 상품명
    - category : 카테고리명으로 검색 시 => keyword : 카테고리명
    - character : 캐릭터명으로 검색 시 => keyword : 캐릭터명
- keyword : 필터링 타입에 따른 검색 키워드
- pageSize : 페이지 사이즈 한번 요청시 받아올 결과 갯수
- page : 현재 페이지
- pageSize 와 page를 포함 시키지 않으면 전체 결과를 반환함

요청 결과 예시(백 -> 프)
```json
{
    "resultList": [
        {
            "id": 90,
            "name": "[마지막기회] 신소재 앵무새 집",
            "price": "82500.000",
            "stock": true,
            "like": false,
            "cart": false,
            "image": "https://i.picsum.photos/id/866/256/256.jpg?ramdom=1"
        },
        {
            "id": 89,
            "name": "[마지막기회] 신소재 고슴도치 집",
            "price": "83100.000",
            "stock": true,
            "like": false,
            "cart": false,
            "image": "https://i.picsum.photos/id/866/256/256.jpg?ramdom=1"
        }
    ],
    "page": 1,
    "pageSize": 2,
    "totalCount": 90,
    "totalPageCount": 45,
    "numberOfElements": 2
}
```

1. resultList : 결과 리스트
    - id : 상품 id값
    - name : 상품 이름
    - price : 상품 가격
    - stock : 재고 여부
    - like : 좋아요 여부
    - cart : 장바구니 여부
    - image : 대표이미지 url

2.  page : 현재 페이지 번호
3. pageSize : 페이지 한개당 결과 갯수
4. totalCount : 전체 아이템 갯수
5. totalPageCount : 전체 페이지 갯수
6. numberOfElements : 결과 리스트의 길이

**page와 pageSize를 쿼리파라미터로 넘기지 않으면 전체 결과를 반환하고, `page`, `pageSize`, `totalPageCount`를 결과에 포함시키지 않도록 함**


![image](https://user-images.githubusercontent.com/8219812/121661569-4b03f380-cadf-11eb-8cad-3defc080c7a5.png)


## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
> **이 내용은 리펙토링 하여 해결하였습니다**
1. like수가 많은 순서대로 product를 얻고 싶어서 group by를 사용하여 count를 이용해서 했습니다.
   - 하지만 무조건 values를 사용해야되서 쿼리셋으로 반환 받지 못하였습니다
   - 그래서 like수가 많은순으로 product id를 list로 받은 후, 스텍 오버플로우 참고하여 조건에 맞게 정렬된 쿼리셋을 가져오도록 하였는데, 다른 방법이 있을 지 궁금합니다.

2. 쿼리스트링 중에 type에 잘못된 값이 들어왔을 때, 빈 list를 반환, 입력을 안 했을시에는 모든 상품을 반환하도록 하였는데, 이렇게 처리하는 것이 괜찮은지 궁금합니다.

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

jotasic committed:
gitigonre에 예외추가 csv DB에 넣는 코드

jotasic committed:
Add: Product 조회 기능 추가

- 쿼리파라미터에 따라서 상품들의 리스트를 응답으로 줄 수 있도록 기능 구현
- 무한스크롤 기능을 대비하여 페이지 번호 및 페이지 사이즈를 받아서
  처리할 수 있도록 하였으며, 해당 파라미터가 없을 시, 키워드에 대한 모든
정보를 응답하도록 함.

예시)
/products?type=타입명&keyword=키워드&pageSize=페이지사이즈&page=페이지번호
- type : 필터링 타입 new,hot,search,category,character가 값으로 올 수
  있음
- keyword : 필터링 타입에 따른 검색 키워드
- pageSize : 페이지 사이즈 한번 요청시 받아올 결과 갯수
- page : 현재 페이지
- pageSize 와 page를 포함 시키지 않으면 전체 결과를 반환함

요청 결과 예시(백 -> 프)
{
    "resultList": [
        {
            "id": 90,
            "name": "[마지막기회] 신소재 앵무새 집",
            "price": "82500.000",
            "stock": true,
            "like": false,
            "cart": false,
            "image": "https://i.picsum.photos/id/866/256/256.jpg?ramdom=1"
        },
        {
            "id": 89,
            "name": "[마지막기회] 신소재 고슴도치 집",
            "price": "83100.000",
            "stock": true,
            "like": false,
            "cart": false,
            "image": "https://i.picsum.photos/id/866/256/256.jpg?ramdom=1"
        }
    ],
    "page": 1,
    "pageSize": 2,
    "totalCount": 90,
    "totalPageCount": 45,
    "numberOfElements": 2
}

1. resultList : 결과 리스트
    - id : 상품 id값
    - name : 상품 이름
    - price : 상품 가격
    - stock : 재고 여부
    - like : 좋아요 여부
    - cart : 장바구니 여부
    - image : 대표이미지 url

2.  page : 현재 페이지 번호
3. pageSize : 페이지 한개당 결과 갯수
4. totalCount : 전체 아이템 갯수
5. totalPageCount : 전체 페이지 갯수
6. numberOfElements : 결과 리스트의 길이

jotasic committed:
Modify: 머지 충돌 해결

jotasic committed:
Add: ProductView에 로그인 토큰값 적용

jotasic committed:
Refactor: 인기 상품순으로 정렬된 결과 반환로직 수정

기존에 정참조를 통해서 id순으로 list를 얻어서 다시 product를 검색했던
복잡한 과정을 역참조를 이용해서 간단하게 표현할 수 있도록 수정함.

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,71 @@
+from users.utils                  import login_decorator
+
+class ProductView(View):
+    @login_decorator
```
좋아요 때문에 로그인 데코레이터를 사용하신 건가요?

jotasic replied:
저희가 로그인이 안되었을 때, main페이지로 진입이 안되게 하도록 하였기 때문에, 로그인 데코레이터 사용하였습니다.

jotasic replied:
이 부분은 저희 팀원분들과 다시 말해봤는데, 이대로 진행될 것 같습니다.

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,71 @@
+        user         = request.user
+        product_type = request.GET.get('type', '')
+        keyword      = request.GET.get('keyword', '')
+        page         = int(request.GET.get('page', 0))
+        page_size    = int(request.GET.get('pageSize', 0))
```
페이지네이션을 위한 offset과 limit로 보입니다. 만약에 페이지네이션을 사용될 값이 프론트에서 안 들어 온다면, default로 넘겨줄 offset과 limit를 미리 정해 주면 아래 로직이 더 간단해 집니다.

jotasic replied:
구두로 말씀하신대로 프론트에서 페지네이션을 사용하지 않아도 기본 값(0)을 받는 것은 이해를 하였는데,
로직이 간단해 진다는 이해를 하지 못한것 같습니다. 지금도 프론트에서 해당값을 주지않으면, 0값으로 사용하기 때문입니다

Kang-kyunghun replied:
전부를 다 줘야 한다는 설명을 듣기 전에 남긴 코멘트 입니다. 전부를 다 줘야하는 부분이 있다면 어쨌든 존재해야 하는 로직인 것은 맞는 것 같아요

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,71 @@
+        page         = int(request.GET.get('page', 0))
+        page_size    = int(request.GET.get('pageSize', 0))
+                
+        if product_type == 'new':
+            filtered_products = Product.objects.order_by('-created_at')
+
```
필터링 부분은 저랑 구두로 한 번 이야기 할게요~

jotasic replied:
필터링 말씀하신 사항 1차 수정 완료하였습니다.

jotasic committed:
Refactor: 조건에 맞게 상품목록을 검색하는 부분 수정

- 기존 : type에 맞게 if문 안에서 일일이 filter를 이용해서 쿼리셋을
  받아냄
- 변경 : Q객체, List, dic를 이용해서 검색할 조건들을 취합하고 마지막에
  하나의 filter문으로 처리되도록 수정함.

jotasic committed:
Fix: 충돌해결

jotasic committed:
Modify: OrderList에서 OrderItem으로 모델명 변겨으로 인한 파라미터 수정

jotasic committed:
Modify: OrderStatus 기존 string형태에서 enum형태로 변경

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,78 @@
+        page_size           = int(request.GET.get('pageSize', 0))
+        order_conditions    = []
+        annotate_conditions = {}
+        filter_conditions   = Q()
```
```suggestion
        q   = Q()
```
Q()을 사용할 때 위와 같이 소문자 q에 Q()를 담는 것이 암묵적인 방법입니다. 당연히 어떤 변수에 넣어도 상관 없지만, 협업하는 입장에서 장고에서 암묵적으로 가져가는 룰은 지키는 것이 좋다고 생각합니다.

jotasic replied:
수정 완료했습니다

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,78 @@
+            'price' : product.price,
+            'stock' : product.stock,
+            'like'  : product.user_set.filter(id=user.id).exists(),
+            'cart'  : product.orderitem_set.all().filter(
+                order__user__id=user.id,
+                order__order_status__status=OrderStatus.BASKET).exists(),
```
```suggestion
            'cart'  : product.orderitem_set.filter(
                order__user = user,
                order__order_status__status=OrderStatus.BASKET).exists(),
```
- filter()을 사용할 경우 앞에 all()을 사용하지 않아도 됩니다.
- user는 request로부터 얻기 때문에 굳이 id까지 안 가도 될 것 같습니다.

jotasic replied:
수정완료하였습니다.

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,78 @@
+            'cart'  : product.orderitem_set.all().filter(
+                order__user__id=user.id,
+                order__order_status__status=OrderStatus.BASKET).exists(),
+            'image' : product.imageurl_set.order_by('id')[0].url
```
이미지는 무조건 하나만 존재하나요? 1장만 보내는 이유가 있을까요?

jotasic replied:
네 제품목록 페이지의 경우 이미지 중 1개만 표시되고,
상세페이지에서는 모든 이미지가 필요합니다.

따라서 제품목록을 보내는 기능인 ProductView 에서는 이미지 중 대표 이미지 1장만 보내고,
상세페이지에서 모든 이미지를 보내고 있습니다.

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,78 @@
+            result['pageSize']       = page_size
+            result['totalPageCount'] = math.ceil(total_count / page_size)
+
+        result['totalCount']       = total_count
+        result['numberOfElements'] = filtered_products.count()
```
2개의 데이터가 같은 것 같은데, 아니라면 설명 코멘트 부탁드립니다.

jotasic replied:
totalCount : 해당 조건으로 검색시 전체 데이터 개수 입니다.
numberOfElements : 반환하는 상품정보에 대한 갯수입니다. 일반적으로 pageSize와 동일하겠지만 마지막 페이지에서는 달라질 수 있습니다.

ex) 최신 상품순으로 검색시 전체 data가 11개가 될 시에 `totalCount`는 11개 입니다.
만약 이 조건에서 `pageSize`가 10 일 경우 2번째 page 에서는 `numberOfElements`는 1이 됩니다.

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,78 @@
+            filter_conditions |= Q(name__icontains=search)
+            order_conditions.append('-created_at')
+        
+        filtered_products = Product.objects.filter(
+            filter_conditions).annotate(
+                **annotate_conditions).order_by(
+                *order_conditions)
```
```suggestion
        filtered_products = Product.objects.filter(q)
                                   .annotate(like_count=Count(like__product_id)
                                   .order_by(-'created_at')
```
1. like count는 조건에 상관없이 annotate를 해도 무방할 것으로 보입니다.
2. order_by 조건이 create_at 밖에 없으므로 마지막에 한 번 사용해 주면 좋을 것 같습니다.

jotasic replied:
> like count는 조건에 상관없이 annotate를 해도 무방할 것으로 보입니다

어짜피 hot일때 말고 count를 사용해서 정렬을 하지 않으니 상관 없군요!

> order_by 조건이 create_at 밖에 없으므로 마지막에 한 번 사용해 주면 좋을 것 같습니다

type의 잘못된 option이 들어올 시에 안하도록 하고 싶었는데, 너무 깊은 생각이였던것 같습니다.!

Kang-kyunghun commented:
중복 필터링이 존재하지 않는다면 현재보다 간단히 바꿀수 있을 것 같은데, 오전에 일찍 같이 코드 보면서 얘기하면 좋을 것 같습니다.

jotasic committed:
Refactor: 상품목록 필터링 코드 수정

1. Q객체를 담는 변수명 암묵적인룰에 맞게 q로 변경
2. 최신순으로 정렬하는 옵션인 -created_at을 모든 조건에서 사용하므로 각
   if문에서 빼고, 공통으로 추가하도록 수정
3. annotate 조건은 hot에서 정렬에 따라서 사용하므로 조건을 변수에 담지
   않고, 바로 사용할 수 있도록 수
4. 그외 불필요한 구문 수정(user_id > user, .all().filter() > filter())

jotasic commented:
> 중복 필터링이 존재하지 않는다면 현재보다 간단히 바꿀수 있을 것 같은데, 오전에 일찍 같이 코드 보면서 얘기하면 좋을 것 같습니다.

jotasic committed:
Fix: order_status 조건 잘못 기입으로 인한 수정

jotasic committed:
Refactor: 검색 필터링 로직 최적화 (진행중)

jotasic committed:
Modify: 병합충돌 해결

jotasic committed:
Modify: 컨벤션 수정

jotasic committed:
Modify: 정렬 조건 직관적으로 수정

jotasic commented:
제가 어제 이야기 한것에 생각을 못하고 맞다고 해서 혼동이 왔습니다. 

 기존에 이렇게 되어 있는 코드를
```python
Product.objects.filter(q).annotate(like_count=Count(like__product_id).order_by(-'created_at')
```

이러한 방식으로 수정하라고 피드백 받았는데 그 당시에는 맞는 것 같았으나, 막상 구현을 하려보니
저 필터의 q.get()을 할 때 쓸 키값을 어디서 구해야될지가 잘 안됩니다.

```python
Product.objects.filter(q.get(type, Q())).annotate(like_count=Count(like__product_id).order_by(-'created_at')
```
또한 어제 생각해보니 구현해야되는 조건 자체는 or이지만 and로 구현하면 나중에 더 유연할 것 같다라는 생각이 드는데,

어떻게 생각하시나요?

이점에 대해서 도움을 드립니다.

jotasic committed:
Modify: 검색 필터링 조건 수정
- 기존에 카테고리, 검색, 캐릭터 중 1개의 검색조건만 적용하려 하였으나,
  구조 변경후에 모든 조건을 and로 하면 더 유연해 질 것 으로 판단되어
수정진행

jotasic committed:
Modify: ProductView 이름 ProductListView로 이름 변경

jotasic commented:
필터링 관련해서 피드백 받은대로 수정하였습니다.

제가 초기에 질문드린 내용은 search, category, character 중 1개의 조건만을 사용해야 되야되는 상황으로 질문드려 피드백을 주셨는데, 피드백 주신 내용을 바탕으로 계속 구조를 바꿔가니 search, category, character의 조건을 and로 하는 것이 기존 기획을 유지하면서 구조적으로 더 유연해진다는 것을 알게되어 좋았습니다.

감사합니다!!

Kang-kyunghun commented:
고생 많으셨습니다~~!!!!!
